contracts/src/subnet/SubnetActorGetterFacet.sol

```diff
@@ -55,8 +55,40 @@ contract SubnetActorGetterFacet {
     }
 
     /// @notice Returns the initial set of validators of the genesis block.
-    function genesisValidators() external view returns (Validator[] memory) {
-        return s.genesisValidators;
+    function genesisValidators() external view returns (GenesisValidator[] memory validators) {
+        uint256 total = s.genesis.validators.length();
+
+        validators = new GenesisValidator[](total);
+
+        if (s.bootstrapped) {
+            // subnet boostrapped, that means validator information at genesis should be locked
+            // and cannot be changed anymore, fetch from s.genesis
+
+            for (uint256 i = 0; i < total; ) {
+                address addr = s.genesis.validators.at(i);
+                validators[i] = s.genesis.validatorInfo[addr];
+
+                unchecked {
+                    i++;
+                }
+            }
+            return validators;
+        }
+
+        for (uint256 i = 0; i < total; ) {
+            address addr = s.genesis.validators.at(i);
+            ValidatorInfo memory info = getValidator(addr);
+            validators[i] = GenesisValidator({
+                // for genesis validators, totalCollateral == confirmedCollateral
+                collateral: info.totalCollateral,
+                federatedPower: info.federatedPower,
```

Comment:
Can these be 0? I can't see where a validator is removed from `SubnetGenesis`, but I assume that they can leave before genesis and remove their collateral?

Reply:
yeah, when they `unstake` or `leave` with zero collateral, they are removed from genesis validator.